src/server.js

```diff
@@ -250,6 +252,91 @@ function moveBranch(repo, branch, sha) {
 			() => repo.git.refs.create({ sha, ref: `refs/${refName}` }));
 }
 
+/**
+ * Cancels current and/or pending builds for all a job associated with a given PR.
+ * Note: The job must define a prId string parameter to support cancellation.
+ */
+function cancelAssociatedCurrentAndPendingBuildsIfNeeded(job, primaryPrId) {
+	const cancelPendingBuilds = superagent
+		.get(url.resolve(job.url, '/queue/api/json'))
+		.then(allQueuedBuildsResponse => {
+			if (allQueuedBuildsResponse && allQueuedBuildsResponse.body && allQueuedBuildsResponse.body.items) {
+				const items = allQueuedBuildsResponse.body.items.filter(x => x && x.task && x.task.name === job.name);
+				return Promise.all(items
+					.map(item => superagent
+						.get(url.resolve(job.url, `/queue/item/${item.id}/api/json`))
+						.then(itemDetailsResponse => {
+							const actions = itemDetailsResponse && itemDetailsResponse.body && itemDetailsResponse.body.actions || [];
+							const matchingBuild = flatMap(actions, x => (x.parameters || [])).some(y => y.name === "prId" && y.value === primaryPrId);
+
+							if (matchingBuild) {
+								const cancelRequest = superagent
+									.post(url.resolve(job.url, `/queue/cancelItem?id=${item.id}`))
+									.auth(jenkinsCredentials.user, jenkinsCredentials.password);
+								return cancelRequest.then(() => {
+									log.info(`Canceled pending build for PR: ${primaryPrId} and Job: ${job.name}`);
```

Comment:
Since we're filtering Jenkins notifications by `COMPLETED`: https://github.com/bgrainger/leeroy-pull-request-builder/blob/cb65720b8d1f52caafab88abf43619e193df523f/src/server.js#L545

... we should add logic here to immediately clean up the branches that were created for the build we're cancelling. Not sure if we have the data at hand, but the code would be something like: https://github.com/bgrainger/leeroy-pull-request-builder/blob/cb65720b8d1f52caafab88abf43619e193df523f/src/server.js#L555